Ipcc patch3

Parisra opened the pull request:
Hi @martavp and @TimToernes ,
 This pull request includes changes that solve some of the current issues such as the filtering method and adding variables like 'added capacity' (I will also answer each issue separately). 
I've added many variables, and the next step is to use the Config file to filter which variables should be calculated for each network. Currently the script writes all the variables to the excel, so even if for example 'industry' is not in the network, all variables related to it will be '0' in the output table, which could be misleading. 
Other comments:
- The variables definitely need some checking. I will try to do this myself and would appreciate any feedback on wether some calculations/numbers/units don't seem correct.
-  I am still trying to find an efficient way to add 'Cumulative capacity' variables to the script.
-  There is some overlap between 'carbon intensity' and 'carbon emissions'. Currently I have only added 'Carbon intensity' since the emissions are divided between specific gases (CO2, Nox, etc.), but there is no variable for equivalent carbon emissions, which I believe is closest to the number calculated by PyPSA.
-  'Agriculture' related variables are not yet added.
-  'Primary energy' is not included yet.

Parisra committed:
Adding new constants and dictionaries for variables

Parisra committed:
Adding new required tables

Minor fix for years, for when it doesn't begin with 2020

Parisra committed:
Updating/adding variables and filtering method

-All variables are updated to use a filtering method that works for multi-node countries
-New variable groups are added : Investment, capacity additions, efficiency, carbon intensity

Parisra committed:
Deleting extra comments

TimToernes commented on pypsa_to_IPCC.py:
```diff
@@ -293,14 +403,108 @@
+                                         n.links.capital_cost.filter(like='coal CC').filter(like=str(year)).filter(like=country)).sum()
+            var['Investment|Energy Supply|Electricity|Coal|w/o CCS'] = (n.links.p_nom_opt.filter(like='coal').filter(like=str(year)).filter(like=country)*
+                                         n.links.capital_cost.filter(like='coal').filter(like=str(year)).filter(like=country)).sum()
+            var['Investment|Energy Supply|Electricity|Electricity Storage'] = (n.links.p_nom_opt.filter(like='coal').filter(like=str(year)).filter(like=country)*
```
I think there is an error here. The investment variable should be energy storage, but you are getting investment in coal

Parisra replied:
Thanks , I fixed this and used batteries for electricity storage

TimToernes commented on pypsa_to_IPCC.py:
```diff
@@ -293,14 +403,108 @@
+            var['Investment|Energy Supply|Electricity|Electricity Storage'] = (n.links.p_nom_opt.filter(like='coal').filter(like=str(year)).filter(like=country)*
+                                         n.links.capital_cost.filter(like='coal').filter(like=str(year)).filter(like=country)).sum()
+            var['Investment|Energy Supply|Electricity|Fossil'] = (n.links.p_nom_opt.filter(like='coal').filter(like=str(year)).filter(like=country)*
+                                         n.links.capital_cost.filter(like='coal').filter(like=str(year)).filter(like=country)).sum()
```
Should the fossil investments only include coal investments?

Parisra replied:
Also fixed. Now includes gas (CCGT and OCGT) and coal, but maybe I'll somehow add CHPs later as well.

TimToernes reviewed (commented):
Looks good. I have two minor comments on the script.

Parisra committed:
Fixes from Tim's comments

comments: https://github.com/martavp/pypsa-eur-sec-to-ipcc/pull/33#pullrequestreview-1251542182

martavp commented on pypsa_to_IPCC.py:
```diff
@@ -10,6 +10,10 @@
 
+MWh2EJ=3.6e-9     #convert MWh to EJ
+TWh2TJ=3.6e+3   #convert TWh to TJ
+MW2GW=0.001
```
this is nice but missing to be implemented in many lines of the code

Parisra replied:
Yes, I changed the list and replaced all the hardcoded numbers like 0.001.

martavp replied:
done in https://github.com/martavp/pypsa-eur-sec-to-ipcc/pull/33/commits/b940cb1a9c69d8ef9dbd2d23f0154a206d8fe54b . Thanks

martavp commented on pypsa_to_IPCC.py:
```diff
@@ -121,136 +121,226 @@
+               """
+               #MW -> GW
+               var[v_type+'|Electricity|Solar|PV'] =0.001*(n.generators.p_nom_opt.filter(like ='solar').filter(like =country).filter(like=str(dict_var[v_type])).sum()-
+                                                          n.generators.p_nom_opt.filter(like ='solar thermal').filter(like =country).filter(like=str(dict_var[v_type])).sum())
```
this is a little counterintuitive, I suggest adding a comment saying #solar thermal is subtracted to only account for solar PV

Parisra replied:
Good point, I added a comment.

martavp replied:
thanks

martavp commented on pypsa_to_IPCC.py:
```diff
@@ -121,136 +121,226 @@
+               Capacity : storage (PHS, battery, H2 storage, thermal storage)
+               """
+               #MWh to GWh
+               var[v_type+'|Electricity|Storage|Pumped Hydro Storage'] = 0.001*(n.storage_units.p_nom_opt.filter(like ='PHS').filter(like =country).filter(like=str(dict_var[v_type])).sum())
```
this reports the power capacity (instead of the intended energy capacity), you need to multiply by the attribute max_hours

the mentioned units are MWh to GWh, so this should be energy capacity, right?

Parisra replied:
I checked and there is a inconsistency in the IPCC table, as most of the storage 'capacities' have the unit 'GWh/year', but the two general variables 'Capacity|Electricity|Storage Capacity' and 'Capacity|Electricity|Storage Capacity'' have GW as the unit. I changed all the storage definitions to match this, but I'm not sure if this was intentional or there is a mistake with the 'GW' units. I also only included one of the 'storage capacity' variables since I can't tell how they're different.

martavp replied:
I think we should keep all the storage reported in GWh/year. 
Is this what you did?

Parisra replied:
Okay, I had kept 1 variable with GW, but deleted it out so that all storage capacities have the GWh unit

martavp commented on pypsa_to_IPCC.py:
```diff
@@ -121,136 +121,226 @@
+               #MWh to GWh
+               var[v_type+'|Electricity|Storage|Pumped Hydro Storage'] = 0.001*(n.storage_units.p_nom_opt.filter(like ='PHS').filter(like =country).filter(like=str(dict_var[v_type])).sum())
+               var[v_type+'|Electricity|Storage|Battery Capacity|Home Battery'] = 0.001*((n.links.efficiency.filter(like ='home battery charger').filter(like =country).filter(like=str(dict_var[v_type]))
+                 *n.links.p_nom_opt.filter(like ='home battery charger').filter(like =country).filter(like=str(dict_var[v_type]))).sum())
```
this reports the power capacity, to report the energy capacity you need to read n.stores.e_nom_opt

Parisra replied:
fixed, above explanation

martavp commented on pypsa_to_IPCC.py:
```diff
@@ -121,136 +121,226 @@
+                 - var[v_type+'|Electricity|Storage|Battery Capacity|Home Battery'])
+               #var[v_type+'|Electricity|Storage|Hydrogen Storage Capacity|overground'] = 0.001 *(n.stores.e_nom_opt.filter(like ='H2').filter(like =country).filter(like=str(dict_var[v_type])).sum()/168) #assume one week charge time for H2 storage
+               #var[v_type+'|Electricity|Storage|Hydrogen Storage Capacity|underground'] = 0.001 *n.stores.e_nom_opt[country + ' H2 Store underground'] if country + ' H2 Store underground' in n.stores.index else 0
+               var[v_type+'|Electricity|Storage|Hydrogen Storage Capacity'] = 0.001 *(n.stores.e_nom_opt.filter(like ='H2').filter(like =country).filter(like=str(dict_var[v_type])).sum()/168)
```
this reports the power capacity, to report the energy capacity no need to divide by 168

also, let's avoid hard-coding numbers, this can be read from the attribute max_hours

Parisra replied:
I removed the 168 for the energy capacity. As for power capacity (added in 'Capacity|Electricity|Storage Capacity'), I'm now using the fuel cell link capacity as the power capacity of hydrogen storage.

Parisra replied:
This is now deleted since it does not match the rest of the storage capacities with unit GWh

martavp commented on pypsa_to_IPCC.py:
```diff
@@ -121,136 +121,226 @@
             """
             #MWh -> EJ
-            var['Secondary Energy|Electricity|Solar|PV'] = 3.6e-9*n.generators_t.p[country + ' solar'].sum() if country + ' solar' in n.generators.index else 0
+            var['Secondary Energy|Electricity|Solar|PV'] = MWh2EJ*h*(n.generators_t.p.filter(like ='solar').filter(like =country).sum().sum())
```
where is the variable h defined?

Parisra replied:
I added a definition  that uses the scenario name from Config, but could be improved.
https://github.com/Parisra/pypsa-eur-sec-to-ipcc/blob/b940cb1a9c69d8ef9dbd2d23f0154a206d8fe54b/pypsa_to_IPCC.py#L30

martavp replied:
perfect, I missed that definition

martavp commented:
I added (in the menu on the left -Development) the number of some of the issues that will be closed when this PR is accepted. Could you check if that is correct and add other issues if relevant?

martavp reviewed (commented):
Thanks, @Parisra I added small suggestions. When you accept them, I can merge the PR and we keep improving it

martavp commented on pypsa_to_IPCC.py:
```diff
@@ -104,14 +104,17 @@
         n = pypsa.Network(f"postnetworks/{scenario}{year}.nc")
         costs = pd.read_csv(f"costs/costs_{year}.csv", index_col=[0,1])
-        
+        prod = pd.read_csv("resources/industrial_production_elec_s_37_{}.csv".format(year),index_col=0,header=0,)  #s_37 ?? 
```
the part in the file name '37' should not be hard-codded but read from the config file

Parisra replied:
Yes it should, fixed it.

martavp replied:
thanks

martavp commented on pypsa_to_IPCC.py:
```diff
@@ -104,14 +104,17 @@
         costs = pd.read_csv(f"costs/costs_{year}.csv", index_col=[0,1])
-        
+        prod = pd.read_csv("resources/industrial_production_elec_s_37_{}.csv".format(year),index_col=0,header=0,)  #s_37 ?? 
+        industry_demand=pd.read_table('resources/industrial_energy_demand_elec_s370_37m_{}.csv'.format(year),delimiter=',',index_col=0)
```
the part in the file name 's370_37m' should not be hard-codded but read from the config file

Parisra replied:
Also fixed.

martavp replied:
fixed in https://github.com/martavp/pypsa-eur-sec-to-ipcc/pull/33/commits/b940cb1a9c69d8ef9dbd2d23f0154a206d8fe54b . Thanks

Parisra commented:
#26

Parisra commented:
#26

Parisra committed:
Additional fixes for variables and tables

fixes from comments on the pull request:
https://github.com/martavp/pypsa-eur-sec-to-ipcc/pull/33

martavp commented:
> 
If this issue is closed by this PR, then you can add it to the right, just below "Successfully merging this pull request may close these issues."

martavp commented:
I think this is ready to merge, as soon as you clarify what you did with the storage capacities
For next PR and commits I suggest you use more explanatory names. 
For example "additional fixes" is difficult to understand, you can all it "fixed definition for storage capacities". Also, split the commits into small commits. For example, "fixed definition of storage capacities", "add comments on solar thermal", "implement MW2GW" . This makes it much easier to review.

For the PR also use descriptive names "Ipcc patch 3" is not giving any information. This can be called "fixing storage capacities and MW2GW factor"

Parisra committed:
Fix storage capacities

- Changed definition for battery/home battery storage capacity so that the unit is GWh
- Deleted var['Capacity|Electricity|Storage Capacity'] because it has the unit GW in the IPCC table, and does not match other storage capacities with GWh unit.

Parisra committed:
Adding 'Cumulative Capacity' variables

Cumulative capacity is the sum of all capacity installed for a technology from the 'start year', including both operational capacities and decommissioned capacities. Cumulative capacity is equal to installed capacity for start year. For every year after start year : Cumulative capacity = (Cumulative capacity of the previous year + capacity additions of the current year)

Parisra committed:
Fix 'list index out of range' error for capacity variables

The IPCC table does not include 'capacity','cumulative capacity', and 'capacity additions' for all technologies. If the script is trying to write a non-existent variable in the table, 'the 'list index out of range' error will be raised. The 'if' statement makes sure a variable is in the table before attempting to write it.

Parisra committed:
Adding emission variables

Parisra committed:
Adding prices of energy

Parisra committed:
Add agriculture demand

Parisra committed:
Change numbers to constant variable

Parisra committed:
Add function

'safe_div' function is used to calculate emission shares for CHPs.

Parisra committed:
Fix typo

industry_name is a variable, not an string

Parisra committed:
Fix if statement

Parisra commented:
This pull request is ready to be merged. I will be doing another PR afterwards for updating the documentation.
